llvm/test/Transforms/InstCombine/commutative-operation-over-selects.ll

```diff
@@ -0,0 +1,1117 @@
+  %ret = call i16 @llvm.umul.fix.sat.i16(i16 %s1, i16 %s0, i32 1)
+  ret i16 %ret
+}
+
+define float @fold_select_fma(i1 %c, float %a, float %b, float %y) {
+; CHECK-LABEL: define float @fold_select_fma(
+; CHECK-SAME: i1 [[C:%.*]], float [[A:%.*]], float [[B:%.*]], float [[Y:%.*]]) {
+; CHECK-NEXT:    [[RET:%.*]] = call float @llvm.fma.f32(float [[B]], float [[A]], float [[Y]])
+; CHECK-NEXT:    ret float [[RET]]
+;
+  %s0 = select i1 %c, float %a, float %b
+  %s1 = select i1 %c, float %b, float %a
+  %ret = call float @llvm.fma.f32(float %s1, float %s0, float %y)
+  ret float %ret
+}
+
+define float @fold_select_fmuladd(i1 %c, float %a, float %b, float %y) {
+; CHECK-LABEL: define float @fold_select_fmuladd(
+; CHECK-SAME: i1 [[C:%.*]], float [[A:%.*]], float [[B:%.*]], float [[Y:%.*]]) {
+; CHECK-NEXT:    [[RET:%.*]] = call float @llvm.fmuladd.f32(float [[B]], float [[A]], float [[Y]])
+; CHECK-NEXT:    ret float [[RET]]
+;
+  %s0 = select i1 %c, float %a, float %b
+  %s1 = select i1 %c, float %b, float %a
+  %ret = call float @llvm.fmuladd.f32(float %s1, float %s0, float %y)
+  ret float %ret
+}
+
+;negative tests:
+
+define i8 @fold_select_unmatch_mul_neg(i1 %c, i1 %c1, i8 %a, i8 %b) {
+; CHECK-LABEL: define i8 @fold_select_unmatch_mul_neg(
+; CHECK-SAME: i1 [[C:%.*]], i1 [[C1:%.*]], i8 [[A:%.*]], i8 [[B:%.*]]) {
+; CHECK-NEXT:    [[S0:%.*]] = select i1 [[C]], i8 [[A]], i8 [[B]]
+; CHECK-NEXT:    [[S1:%.*]] = select i1 [[C1]], i8 [[B]], i8 [[A]]
+; CHECK-NEXT:    [[RET:%.*]] = mul i8 [[S1]], [[S0]]
+; CHECK-NEXT:    ret i8 [[RET]]
+;
+  %s0 = select i1 %c, i8 %a, i8 %b
+  %s1 = select i1 %c1, i8 %b, i8 %a
+  %ret = mul i8 %s1, %s0
+  ret i8 %ret
+}
+
+define i8 @fold_select_mul_neg(i1 %c, i8 %a, i8 %b, i8 %y, i8 %z) {
+; CHECK-LABEL: define i8 @fold_select_mul_neg(
+; CHECK-SAME: i1 [[C:%.*]], i8 [[A:%.*]], i8 [[B:%.*]], i8 [[Y:%.*]], i8 [[Z:%.*]]) {
+; CHECK-NEXT:    [[S0:%.*]] = select i1 [[C]], i8 [[A]], i8 [[B]]
+; CHECK-NEXT:    [[S1:%.*]] = select i1 [[C]], i8 [[Y]], i8 [[Z]]
+; CHECK-NEXT:    [[RET:%.*]] = mul i8 [[S1]], [[S0]]
+; CHECK-NEXT:    ret i8 [[RET]]
+;
+  %s0 = select i1 %c, i8 %a, i8 %b
+  %s1 = select i1 %c, i8 %y, i8 %z
+  %ret = mul i8 %s1, %s0
+  ret i8 %ret
+}
+
+define <2 x i4> @fold_select_mul_vec2_neg(i1 %c, <2 x i4> %a, <2 x i4> %b, <2 x i4> %y, <2 x i4> %z) {
+; CHECK-LABEL: define <2 x i4> @fold_select_mul_vec2_neg(
+; CHECK-SAME: i1 [[C:%.*]], <2 x i4> [[A:%.*]], <2 x i4> [[B:%.*]], <2 x i4> [[Y:%.*]], <2 x i4> [[Z:%.*]]) {
+; CHECK-NEXT:    [[S0:%.*]] = select i1 [[C]], <2 x i4> [[A]], <2 x i4> [[B]]
+; CHECK-NEXT:    [[S1:%.*]] = select i1 [[C]], <2 x i4> [[Y]], <2 x i4> [[Z]]
+; CHECK-NEXT:    [[RET:%.*]] = mul <2 x i4> [[S1]], [[S0]]
+; CHECK-NEXT:    ret <2 x i4> [[RET]]
+;
+  %s0 = select i1 %c, <2 x i4> %a, <2 x i4> %b
+  %s1 = select i1 %c, <2 x i4> %y, <2 x i4> %z
+  %ret = mul <2 x i4> %s1, %s0
+  ret <2 x i4> %ret
+}
+
+define i8 @fold_select_unmatch_add_neg(i1 %c, i1 %c1, i8 %a, i8 %b) {
```

Comment:
Don't repeat negative tests for all operations. Only test one binop and one intrinsic, but not more than that.